src/flightRecorder.cpp

```diff
@@ -72,6 +79,102 @@ struct CpuTimes {
     CpuTime total;
 };
 
+struct ProcessHistory {
+    float prev_cpu_total = 0.0f;
+    u64 prev_timestamp = 0;
+    u64 start_time = 0;
+};
+
+class ProcessSampler {
+  private:
+    static u64 _last_sample_time;
+    static u64 _next_sample_time;
+    static std::unordered_map<int, ProcessHistory> _process_history;
+
+    long _sampling_interval = -1;
+    int _pids[MAX_PROCESSES];
+
+  public:
+    void cleanupProcessHistory(const int pid_count) {
+        const std::unordered_set<int> active_pid_set(_pids, _pids + pid_count);
+        for (auto it = _process_history.begin(); it != _process_history.end();) {
+            if (active_pid_set.count(it->first) == 0) {
+                it = _process_history.erase(it);
+            } else {
+                ++it;
+            }
+        }
+    }
+
+    void enable(const long sampling_interval) {
+        _sampling_interval = sampling_interval;
+        u64 expected_next = _last_sample_time + sampling_interval;
+        if (_next_sample_time != expected_next) {
+            _last_sample_time = 0;
+            _process_history.clear();
+        }
+    }
+
+    bool shouldSample(const u64 wall_time) const {
+        return _sampling_interval > 0 && wall_time >= _next_sample_time;
+    }
+
+    static double getRssUsagePercent(const ProcessInfo& info) {
+        const u64 ram_size = OS::getRamSize();
+        if (ram_size <= 0 || info.vm_rss <= 0) return 0.0;
```

Comment:
`<= 0` looks strange for unsigned values. Can be replaced with `== 0`.
I remember some static analyzer was complaining about that.